Add banner to previous year events redirecting visitors to the latest event

timiwahalahti opened the pull request:
Add notification if a user visits WordCamp site that has a newer site for the same location. Keeps the canonical link functionality for seo purposes.

Also renamed the file to follow better naming, as it made sense to put this functionality into the same file that already contained hints about the latest site for search engines.

This was a bit tricky to test because of the differences between the Docker environment and production. Also, the development database is lacking proper real-world examples. So needs proper thorough testing, mentioning @iandunn here as they had written the seo code on which this PR is based.

Fixes #688

Props @dd32 

### Screenshots

<img width="1164" alt="image" src="https://user-images.githubusercontent.com/415544/148633354-6b83dfc9-9cd8-4913-8059-b861ef765bdc.png">

### How to test the changes in this Pull Request:

1. Create new WordCamp entry on tracker, based on some other existing event but change the year
2. Make sure the new WordCamp has a site
3. Visit the WordCamp site from which you based the new event, notification should be visible

timiwahalahti committed:
Add notification if user visits WordCamp site that has newer site for the same location. Rename file to follow better naming.

timiwahalahti committed:
Fix tests bootstrapping

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+	 */
+	if ( ! is_user_logged_in() ) {
+		add_filter( 'body_class',   __NAMESPACE__ . '\add_notification_classes_to_body' );
+		add_action( 'wp_head',      __NAMESPACE__ . '\add_notification_styles' );
```
I'm seeing this fire when I'm logged in, and not seeing it when logged out. I don't see `show_notification_about_latest_site()` fire when logged out either.

Maybe it's something wrong in my env? I can dig deeper if this is still working in yours.

timiwahalahti replied:
Hmm, I couldn't reproduce this. Non-logged-in users do see the the banner and get inline styles, logged-in users do not. Of course, the `admin-bar` body class is there for logged-in users via core.

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+
+/**
+ * If notification is shown, misuse body.admin-bar class as most of the WordCamp themes already that that into
+ * account. This makes its a whole lot easier to place the notification on top of contens.
```
```suggestion
 * account. This makes its a whole lot easier to place the notification on top of content.
```

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+function add_notification_styles() { ?>
+	<style type="text/css">
+		html {
+			margin-top: 35px !important;
```
```suggestion
			margin-top: var(--wp-admin--admin-bar--height) !important;
```

That was recently added to Core, and will automatically update from `32px` to `46px` at the various breakpoints.

timiwahalahti replied:
The idea is good, although the `--wp-admin--admin-bar--height` is defined in `admin-bar.css`, which isn't enqueued in the public site when user is logged out. If we end up loading that stylesheet, then the use of that variable is definitely wise.

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+			text-align: center;
+			padding-top: 5px;
+			font-size: 15px;
+			height: 35px;
```
This might also need to use `--wp-admin--admin-bar--height`, if it's intended to match the `html margin-top`

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+}
+
+/**
+ * If notification is shown, misuse body.admin-bar class as most of the WordCamp themes already that that into
```
That's a clever idea, kudos 👍🏻

iandunn commented on public_html/wp-content/mu-plugins/tests/bootstrap.php:
```diff
@@ -27,7 +27,7 @@ function manually_load_plugins() {
 	require_once dirname( __DIR__ ) . '/0-error-handling.php';
 	require_once dirname( __DIR__ ) . '/lets-encrypt-helper.php';
-	require_once dirname( __DIR__ ) . '/seo.php';
+	require_once dirname( __DIR__ ) . '/latest-site-hints.php';
```
It looks like the [tests are failing](https://github.com/WordPress/wordcamp.org/runs/4746639950?check_suite_focus=true) because the cases need to be updated to reflect the changes in `get_latest_site_id()`. 

`test:watch -- --group=latest-site-hints` makes it convenient. [The Docker readme](https://github.com/WordPress/wordcamp.org/blob/f58e6fc89a5cb755574576b5e0ab935c57449e51/.docker/readme.md) has more details.

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+
+	$latest_site = get_site( $latest_site_id );
+
+	if ( $latest_site->domain === $current_blog->domain ) {
```
This needs to check the `path` as well, since the year is now there. It's probably easier to compare the `blog_id` though.

iandunn commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+			font-size: 15px;
+			height: 35px;
+			position: fixed;
+			top: 0;
```
Why are are the positioning styles needed if we're reusing the admin bar classes? I'd assume it'd inherit those. Maybe enqueuing the stylesheet would add too much unnecessary stuff or something?

timiwahalahti replied:
I think the original idea was to avoid loading the admin-bar stylesheet and make things lighter since not all admin bar styles work well with this notification and the admin bar stylesheet `html` tag `scroll-padding-top` does not seem to work at least with Firefox.

iandunn reviewed (commented):
Sorry for the delay here :(

I went ahead and merged the file renaming bits, to make it easier to review the actual changes.

I think this is good overall, just need to figure out a possible bug and a few details.

ryelle commented on public_html/wp-content/mu-plugins/latest-site-hints.php:
```diff
@@ -0,0 +1,216 @@
+	if ( ! is_user_logged_in() ) {
+		add_filter( 'body_class',   __NAMESPACE__ . '\add_notification_classes_to_body' );
+		add_action( 'wp_head',      __NAMESPACE__ . '\add_notification_styles' );
+		add_action( 'wp_body_open', __NAMESPACE__ . '\show_notification_about_latest_site' );
```
This hook is only called if the theme uses [the `wp_body_open` function](https://make.wordpress.org/core/2019/04/24/miscellaneous-developer-updates-in-5-2/), and CampSite 2017 doesn't — if there isn't another suitable hook, the theme can be updated.

timiwahalahti replied:
Oh, good catch! Changed the message output to `wp_footer`, but I'm not too thrilled about that because of accessibility. 🤔

dd32 replied:
> This hook is only called if the theme uses [the wp_body_open function](https://make.wordpress.org/core/2019/04/24/miscellaneous-developer-updates-in-5-2/), and CampSite 2017 doesn't

Regardless, we should probably add that to the theme just for future compatibility.

timiwahalahti replied:
Agreed. I will open a PR on the theme, and let's switch to `wp_body_open` when that is in production.

timiwahalahti committed:
merge production and refresh the code

timiwahalahti committed:
Update public_html/wp-content/mu-plugins/latest-site-hints.php

Co-authored-by: Ian Dunn <ian@iandunn.name>

timiwahalahti committed:
code styiling

timiwahalahti commented:
Sorry for the delay! Refreshed the code and made changes based on the review comments. I also made the message translatable and simplified it a bit by removing the site name of the next edition - I think that caused unnecessary queries since we already have the URL for the next edition.

Was thinking of extracting the year from URL, now that it's a path, but that started feeling a bit more fragile - and the lack of year does not even make message much different.

Also removed the europe.wordcamp.org exception since that seems to be covered upper in the same code.

Oh and URL structure change to year as a path seemed to actually broke some checks, since some functions were returning different url/domain than the one we checked against.

StevenDufresne commented:
@timiwahalahti Is there anything specific left that you'd like a review on? It would be nice to get this PR in ahead of WCUS!

timiwahalahti commented:
Nothing special, just re-review since there are changes made after @iandunn and @ryelle last time reviewed :)

StevenDufresne commented:
> Nothing special, just re-review since there are changes made after @iandunn and @ryelle last time reviewed :)

There seems to be a merge conflict, do you mind resolving?

timiwahalahti committed:
resolve merge conflict

timiwahalahti commented:
Sure thing, done!

ryelle committed:
Show notification for logged in users too, remove admin-bar CSS hack

ryelle reviewed (approved):
This is looking good! I do think this notification should appear for logged in users too, so I've just pushed an update to this PR to do that and tweak some of the CSS to better support that.

It's a little janky on small screens, but once it's moved to `wp_body_open` that should be fixable. I think this is good to merge 👍🏻

@renintw I know you were tagged here, are you able to merge and deploy this?

outdoor2kode commented:
Looks great code-wise and the UI displays well!

It's been deployed. But a minor issue was spotted with the admin bar dropdown. On some sites, it gets covered by the banner. (Originally, I thought only the admin account would get this problem. Because locally, when logging in with a non-admin, the dropdown isn't covered by the admin bar. However, in the production env, it gets covered)

![Screen Shot 2023-08-10 at 10 25 24 PM](https://github.com/WordPress/wordcamp.org/assets/18050944/40355b50-5562-47b3-8d6c-c761f1ba7aac)
![Screen Shot 2023-08-10 at 10 29 47 PM](https://github.com/WordPress/wordcamp.org/assets/18050944/46945bb4-c817-4998-8e6f-c55d4f40a4d4)
![Screen Shot 2023-08-10 at 10 24 12 PM](https://github.com/WordPress/wordcamp.org/assets/18050944/3f5590a0-07d3-4130-b031-d38e48c293bc)

I have proceeded with the merge and deployment for this PR, and I'll look into the issue. Seems related to the theme used.

outdoor2kode commented:
@ryelle After testing, I've found that the issue occurs with most themes (2016-2023 at least) except for "CampSite 2017". On sites using themes other than "CampSite 2017", the `div#wpadminbar` is loaded earlier, causing the dropdown to be covered by the later loaded `div.wordcamp-latest-site-notify`.

I've tried the following two solutions and both seem to fix the issue:

1. Adjusting the z-index [here](https://github.com/WordPress/wordcamp.org/pull/725/files#diff-22b591d047311fc622c1be6e8dc4e422ce30d0f0375183b876e2da7d0fdee1d0R84) to 99998.
2. Changing wp_footer to wp_head [here](https://github.com/WordPress/wordcamp.org/pull/725/files#diff-22b591d047311fc622c1be6e8dc4e422ce30d0f0375183b876e2da7d0fdee1d0R30).

But neither option seems like best practice, so I thought I'd get your opinion on how we usually handle this.

StevenDufresne commented:
> Adjusting the z-index [here](https://github.com/WordPress/wordcamp.org/pull/725/files#diff-22b591d047311fc622c1be6e8dc4e422ce30d0f0375183b876e2da7d0fdee1d0R84) to 99998.

That approach seems fine to me, unless there's another reason why its exactly set to the same `z-index` as the adminbar.

outdoor2kode commented:
@StevenDufresne Thanks for the input 🙂 The issue was resolved after deploying the [commit](https://github.com/WordPress/wordcamp.org/commit/d5de714a7396fd2e944db477b34a5af4102775d8).
